Doesn't compile with C++17

Compiling with C++17 fails and gives the following errors. C++20 works fine.
Compiler versions tested: AppleClang 15.0.0.15000309 and Clang 18.1.6
```
[1/23] Building CXX object CMakeFiles/SeQuant.dir/SeQuant/core/context.cpp.o
FAILED: CMakeFiles/SeQuant.dir/SeQuant/core/context.cpp.o 
/opt/homebrew/bin/ccache /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang++ -DBOOST_CHRONO_DYN_LINK -DBOOST_CHRONO_NO_LIB -DBOOST_LOCALE_DYN_LINK -DBOOST_LOCALE_NO_LIB -DBOOST_REGEX_DYN_LINK -DBOOST_REGEX_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DSEQUANT_HAS_EIGEN=1 -DSEQUANT_USE_SYSTEM_BOOST_HASH=1 -I/Users/ajay/Code/SeQuant -I/Users/ajay/Code/SeQuant/cmake-build-debug -I/Users/ajay/Code/SeQuant/cmake-build-debug/_deps/rangev3-src/include -isystem /opt/homebrew/Cellar/boost/1.85.0/include -isystem /opt/homebrew/include/eigen3 -g -std=c++17 -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.5.sdk -fcolor-diagnostics -MD -MT CMakeFiles/SeQuant.dir/SeQuant/core/context.cpp.o -MF CMakeFiles/SeQuant.dir/SeQuant/core/context.cpp.o.d -o CMakeFiles/SeQuant.dir/SeQuant/core/context.cpp.o -c /Users/ajay/Code/SeQuant/SeQuant/core/context.cpp
In file included from /Users/ajay/Code/SeQuant/SeQuant/core/context.cpp:2:
In file included from /Users/ajay/Code/SeQuant/SeQuant/core/context.hpp:5:
In file included from /Users/ajay/Code/SeQuant/SeQuant/core/index_space_registry.hpp:8:
In file included from /Users/ajay/Code/SeQuant/SeQuant/core/space.hpp:14:
In file included from /Users/ajay/Code/SeQuant/SeQuant/core/utility/string.hpp:8:
/Users/ajay/Code/SeQuant/SeQuant/core/meta.hpp:129:16: error: use of undeclared identifier 'char8_t'; did you mean 'char16_t'?
struct is_char<char8_t> : std::true_type {};
               ^~~~~~~
               char16_t
/Users/ajay/Code/SeQuant/SeQuant/core/meta.hpp:129:23: error: expected '(' for function-style cast or type construction
struct is_char<char8_t> : std::true_type {};
               ~~~~~~~^
In file included from /Users/ajay/Code/SeQuant/SeQuant/core/context.cpp:2:
In file included from /Users/ajay/Code/SeQuant/SeQuant/core/context.hpp:5:
/Users/ajay/Code/SeQuant/SeQuant/core/index_space_registry.hpp:1191:26: error: no member named 'contains' in 'std::map<sequant::QuantumNumbersAttr, sequant::TypeAttr>'
      if (!qn2type_found.contains(qn)) {
           ~~~~~~~~~~~~~ ^
3 errors generated.
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1,9 +1,20 @@
 cmake_minimum_required(VERSION 3.15.0) # need list(PREPEND for toolchains
 
+# Preload versions/tags of all dependencies ====================================
+include(external/versions.cmake)
+
+###############################################################################
+# CMake defaults to address key pain points
+###############################################################################
+
+# safety net for dev workflow: accidental install will not affect FindOrFetch*
+if (NOT DEFINED CACHE{CMAKE_FIND_NO_INSTALL_PREFIX})
+    set(CMAKE_FIND_NO_INSTALL_PREFIX ON CACHE BOOL "Whether find_* commands will search CMAKE_INSTALL_PREFIX and CMAKE_STAGING_PREFIX; see https://cmake.org/cmake/help/latest/variable/CMAKE_FIND_NO_INSTALL_PREFIX.html#variable:CMAKE_FIND_NO_INSTALL_PREFIX")
+endif()
+
 ###############################################################################
 # Bring ValeevGroup cmake toolkit
 ###############################################################################
-include(external/versions.cmake)
 include(FetchContent)
 if (DEFINED PROJECT_BINARY_DIR)
     set(VG_CMAKE_KIT_PREFIX_DIR PROJECT_BINARY_DIR)

--- a/SeQuant/core/index.hpp
+++ b/SeQuant/core/index.hpp
@@ -336,7 +336,7 @@ class Index : public Taggable {
       return {label, {}};
     else
       return {{label.data(), underscore_position},
-              {label.begin() + underscore_position + 1, label.end()}};
+              {label.begin() + underscore_position + 1}};
   }
 
   /// @param base_label base part of an Index label

--- a/SeQuant/core/index_space_registry.hpp
+++ b/SeQuant/core/index_space_registry.hpp
@@ -1188,7 +1188,11 @@ class IndexSpaceRegistry {
 
     std::string errmsg;
     for (auto&& [qn, t] : qn2type) {
+#if __cplusplus < 202002L
+      if (qn2type_found.find(qn) == qn2type_found.end()) {
+#else
       if (!qn2type_found.contains(qn)) {
+#endif
         errmsg +=
             call_context +
             ": missing { IndexSpace::Type=" + std::to_string(t.to_int32()) +

--- a/SeQuant/core/meta.hpp
+++ b/SeQuant/core/meta.hpp
@@ -125,8 +125,10 @@ template <>
 struct is_char<char> : std::true_type {};
 template <>
 struct is_char<wchar_t> : std::true_type {};
+#if __cplusplus >= 202002L
 template <>
 struct is_char<char8_t> : std::true_type {};
+#endif
 template <>
 struct is_char<char16_t> : std::true_type {};
 template <>

--- a/examples/synopsis/synopsis6.cpp
+++ b/examples/synopsis/synopsis6.cpp
@@ -5,7 +5,13 @@
 #include <SeQuant/domain/mbpt/convention.hpp>
 #include <SeQuant/domain/mbpt/op.hpp>
 
+#if __cplusplus >= 202002L
 inline auto commutator(auto op1, auto op2) { return op1 * op2 - op2 * op1; }
+#else
+inline auto commutator(sequant::ExprPtr op1, sequant::ExprPtr op2) {
+  return op1 * op2 - op2 * op1;
+}
+#endif
 
 int main() {
   using namespace sequant;

